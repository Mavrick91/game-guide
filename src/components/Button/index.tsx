import { cva, type VariantProps } from 'class-variance-authority'
import { type ButtonHTMLAttributes, type ReactElement } from 'react'
import { cn } from '../../utils/style'

const buttonVariants = cva('flex items-center justify-center font-poppins transition-all, duration-200', {
  variants: {
    variant: {
      ghost: ['bg-transparent'],
      black: ['text-white', 'rounded-xl', 'bg-black', 'active:bg-[#0A0A0A]', 'hover:bg-[#141414]']
    },
    size: {
      small: ['text-sm', 'py-1', 'px-2'],
      medium: ['text-base', 'py-2', 'px-7', 'font-semibold']
    }
  },
  defaultVariants: {
    variant: 'ghost'
  }
})

interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
  VariantProps<typeof buttonVariants> {}

export default function BlackButton({
  className,
  variant,
  size,
  ...buttonProps
}: ButtonProps): ReactElement {
  return (
    <button
      {...buttonProps}
      className={cn(buttonVariants({ variant, size, className }))}
    />
  )
}
