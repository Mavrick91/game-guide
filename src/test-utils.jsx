import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const AllTheProviders = ({ children }) => {
  return children;
};

const customRender = (ui, options) => {
  return {
    user: userEvent.setup(),
    ...render(ui, { wrapper: AllTheProviders, ...options }),
  };
};

export * from '@testing-library/react';

export { customRender as render };
