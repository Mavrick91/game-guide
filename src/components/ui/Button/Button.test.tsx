import Button from '.';
import { render } from '../../../test-utils';

describe('Button', () => {
  it('should render the button', () => {
    const { container } = render(<Button />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with default props', () => {
    const { getByRole } = render(<Button />);
    const button = getByRole('button');

    expect(button).toBeInTheDocument();
  });

  it('handles click events', async () => {
    const handleClick = jest.fn();
    const { getByRole, user } = render(<Button onClick={handleClick} />);
    const button = getByRole('button');

    await user.click(button);

    expect(handleClick).toHaveBeenCalled();
  });

  it('applies custom classes', () => {
    const { getByRole } = render(<Button className='custom-class' />);
    const button = getByRole('button');

    expect(button).toHaveClass('custom-class');
  });

  it('applies the "ghost" variant classes', () => {
    const { getByRole } = render(<Button variant='ghost' />);
    const button = getByRole('button');

    expect(button).toHaveClass('bg-transparent');
    expect(button).toHaveClass('w-full');
    expect(button).toHaveClass('justify-start');
  });

  it('applies the "black" variant classes', () => {
    const { getByRole } = render(<Button variant='black' />);
    const button = getByRole('button');

    expect(button).toHaveClass('text-white');
    expect(button).toHaveClass('rounded-xl');
    expect(button).toHaveClass('bg-black');
    expect(button).toHaveClass('active:bg-[#0A0A0A]');
    expect(button).toHaveClass('hover:bg-[#141414]');
  });

  it('applies the "small" size classes', () => {
    const { getByRole } = render(<Button size='small' />);
    const button = getByRole('button');

    expect(button).toHaveClass('text-sm');
    expect(button).toHaveClass('py-1');
    expect(button).toHaveClass('px-2');
  });

  it('applies the "medium" size classes', () => {
    const { getByRole } = render(<Button size='medium' />);
    const button = getByRole('button');

    expect(button).toHaveClass('text-base');
    expect(button).toHaveClass('py-2');
    expect(button).toHaveClass('px-7');
    expect(button).toHaveClass('font-semibold');
  });

  it('applies the "active" classes', () => {
    const { getByRole } = render(<Button active />);
    const button = getByRole('button');

    expect(button).toHaveClass('text-purple-pop');
  });

  it('applies the "inactive" classes', () => {
    const { getByRole } = render(<Button active={false} />);
    const button = getByRole('button');

    expect(button).toHaveClass('opacity-50');
  });
});
