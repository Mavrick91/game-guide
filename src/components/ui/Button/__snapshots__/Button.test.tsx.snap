// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render the button 1`] = `
<button
  class="flex items-center transition-all, duration-200 gap-3 whitespace-nowrap bg-transparent w-full justify-start"
/>
`;
